### PR TITLE
Fixed the cleanup of non-durable cursors when a reader is closed

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

The non-durable cursors associated with readers are not being cleaned up from the managed ledger. This causes these objects to keep hanging around in memory after the reader is gone.

### CODE REVIEW

**Comment by jiazhai** [comment]
run java8 tests

**Comment by jiazhai** [comment]
run java8 tests

**Comment by jiazhai** [comment]
@merlimat @sijie Could we move this in to 2.5.0 release?

**Comment by jiazhai** [comment]
Thanks @sijie 2.4.2 now

**Comment by sijie** [comment]
retest this please

**Comment by sijie** [comment]
retest this please

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -903,8 +903,9 @@ public void operationComplete() {
                                 ledger.getName(), newPosition, name);
                     }
                 }
-                callback.resetComplete(newPosition);
 
+                callback.resetComplete(newPosition);
+                notifyEntriesAvailable();
```
Comment (ivankelly) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: Seems unrelated.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java [hunk 2]
```diff
@@ -621,70 +627,46 @@ private void resetCursor(Position finalPosition, CompletableFuture<Void> future)
...
-        final CompletableFuture<Void> disconnectFuture;
-
-        // Lock the Subscription object before locking the Dispatcher object to avoid deadlocks
-        synchronized (this) {
-            if (dispatcher != null && dispatcher.isConsumerConnected()) {
-                disconnectFuture = dispatcher.disconnectAllConsumers();
-            } else {
-                disconnectFuture = CompletableFuture.completedFuture(null);
-            }
+        if (dispatcher != null) {
```
[the remark this answers is not in the document]
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java]: Mostly, I started pulling the thread and changing the seek behavior to avoid closing consumers in between. I'll pull that change out (in different PR) and apply it before this PR.

**Comment by wolfstudy** [comment]
ping @merlimat please rebase master code

**Comment by wolfstudy** [comment]
ping @merlimat again,  please rebase master code, thanks.

**Comment by sijie** [comment]
@merlimat can you rebase this pull request? so it can unblock 2.4.2 release. or we can move it to 2.4.3.

**Comment by jiazhai** [comment]
will rebase it

**Comment by jiazhai** [comment]
Hi @merlimat this is handled in #5719, could this be closed?

**Comment by jiazhai** [comment]
Would like to close this first, if needed we could open it
